sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StructuredMessageDecoder.java

```diff
@@ -0,0 +1,182 @@
+package com.azure.storage.common.implementation;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.HashMap;
+import java.util.Map;
+
+public class StructuredMessageDecoder {
+    private static final int DEFAULT_MESSAGE_VERSION = 1;
+    private static final int V1_HEADER_LENGTH = 13;
+    private static final int V1_SEGMENT_HEADER_LENGTH = 10;
+    private static final int CRC64_LENGTH = 8;
+
+    private int messageVersion;
+    private int messageLength;
+    private StructuredMessageFlags flags;
+    private int numSegments;
+    private final ByteBuffer buffer;
+
+    private int messageOffset = 0;
+    private int currentSegmentNumber = 0;
+    private int currentSegmentContentLength = 0;
+    private int currentSegmentContentOffset = 0;
+
+    private long messageCrc64 = 0;
+    private long segmentCrc64 = 0;
+    private final Map<Integer, Long> segmentCrcs = new HashMap<>();
+
+    public StructuredMessageDecoder(ByteBuffer inputBuffer) {
+        this.buffer = inputBuffer.order(ByteOrder.LITTLE_ENDIAN);
+        readMessageHeader();
+    }
+
+    private void readMessageHeader() {
+        if (buffer.remaining() < V1_HEADER_LENGTH) {
+            throw new IllegalArgumentException("Content not long enough to contain a valid message header.");
+        }
+
+        messageVersion = Byte.toUnsignedInt(buffer.get());
+        if (messageVersion != DEFAULT_MESSAGE_VERSION) {
+            throw new IllegalArgumentException("Unsupported structured message version: " + messageVersion);
+        }
+
+        messageLength = (int) buffer.getLong();
```

Comment:
We should also do validation that the message length pulled from the message matches what was passed in as part of the constructor. We should be able to match Python here as well: https://github.com/Azure/azure-sdk-for-python/blob/b896d5c9ee202bd7c3cba51eb602562d41f4d804/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/streams.py#L559-L561